General todo parser.

criske opened the pull request:
Todo parser supports now any language and can parse any text file.
In the case of multiline todos, the algorithm has one drawback : it assumes that user is using 4 spaces for `tab`. So if `tab` is 2 or 8 spaces, line body might misalign with todo starting position (`@`).

criske committed:
General todo parser.

charlesmike commented:
@criske thank you for your Pull Request. I'll assign someone to review it soon.

charlesmike commented:
@amihaiemil I couldn't find any assignee for this task. This is either because there are no contributors with role ``REV`` available or because the project does not have enough funds.

Please, make sure there is at least one available contributor with the required role and the project can afford to pay them.

criske commented on src/test/resources/PhonyClass.java:
```diff
@@ -27,7 +27,7 @@ public void method2() {
      */
     public void method3() {
-        "// @todo #187:30min this should not trigger the todo parser because this is not an actual todo".trim();
+        "// @\\todo #187:30min this should not trigger the todo parser because this is not an actual todo".trim();
```
since the parser is language agnostic, this case will fail, cause is not recognizing `trim()`. Test is fixed by "escaping" the todo

criske commented:
@amihaiemil btw, just in case this pr will be merged, no bounty is necessary :D

criske committed:
General todo parser.

- memory improvement: whole file is not loaded in memory anymore, but is processed using a reader instead.

amihaiemil commented:
@criske well, I can't say I really understand the code. But I didn't understand it before either, and I see the tests are still passing, so it's alright.

How much time did you spend?

criske commented:
```java
/**@todo #1:30min This is a
 *  multiline Todo.
 * This is a simple comment.     
 * @todo #1:30min This is a single line todo.
 */ 
```
Let take this example:
- at lineIndex 0:
   todo starts at index 3 (`@`) : this means `todoPosition = 3` and create Todo header (id, estimation), todo start (line 1), add the first todo line to body `StringBuilder` (`This is a`). 
- at lineIndex 1:
  the todo is started already (`todoPosition = 3`) so we check if character at index 3 (under ` @`) is `space`. In this case it is and we `line.substring(3)` to StringBuilder (` multiline Todo.`).
- at lineIndex 2:
  `todoPosition = 3` but this time the character at index 3 is `T` and not a `space`. This means that todo has ended. We build the todo, setting the end line and adding it to todos. Then we flush the body `StringBuilder` (setting len to 0), and reset `todoPosition` to be -1. But we need to check if this line starts another `todo`. (we could have consecutive `todos`). In our case it isn't and `todoPosition` remains -1.
- at lineIndex 3:
   todo starts at index 1 (`@`) : this means todoPosition = 1 and create Todo header (id, estimation)  todo start (line 4), add the add the first todo line ("This is a") to body `StringBuilder` (`This is a single line todo`). 
- at lineIndex 4:
   the todo is started already (`todoPosition = 1`) but this time line length is 1 (`*` - index 0) so lesser than `todoPosition`. This means again that todo has ended and processes is the same as in `lineIndex` 2 (create the todo and add it list, flush buffer, todoPosition  = -1).
- at lineIndex 5:
  todoPosition is -1, so we check for todo but is none here.

Overall it didn't take long maybe 1:30 hour with all the refactoring.

amihaiemil commented:
@rultor merge it please

rultor commented:
> @rultor merge it please

@amihaiemil OK, I'll try to merge now. You can check the progress of the merge [here](https://www.rultor.com/t/23914-751487737)

rultor commented:
> @rultor merge it please

@amihaiemil Done! FYI, the full log is [here](https://www.rultor.com/t/23914-751487737) (took me 2min)
